File: swift-bridge/Sources/CoreAnimationBridge/CADisplayLink.swift
import AppKit
import Foundation
import QuartzCore

@available(macOS 14.0, *)
@objc final class CAQuartzDisplayLinkTarget: NSObject {
    @objc func tick(_ sender: CADisplayLink) {}
}

@available(macOS 14.0, *)
final class CAQuartzDisplayLinkBox {
    let target: CAQuartzDisplayLinkTarget
    let link: CADisplayLink

    init(target: CAQuartzDisplayLinkTarget, link: CADisplayLink) {
        self.target = target
        self.link = link
    }
}

@available(macOS 14.0, *)
private func caBorrowQuartzDisplayLinkBox(_ handle: UnsafeMutableRawPointer?) -> CAQuartzDisplayLinkBox? {
    guard let handle else { return nil }
    return Unmanaged<CAQuartzDisplayLinkBox>.fromOpaque(handle).takeUnretainedValue()
}

@_cdecl("ca_quartz_display_link_new_main_screen")
public func ca_quartz_display_link_new_main_screen() -> UnsafeMutableRawPointer? {
    guard #available(macOS 14.0, *), Thread.isMainThread else { return nil }
    guard let screen = NSScreen.screens.first ?? NSScreen.main else { return nil }
    let target = CAQuartzDisplayLinkTarget()
    let link = screen.displayLink(target: target, selector: #selector(CAQuartzDisplayLinkTarget.tick(_:)))
    return caRetain(CAQuartzDisplayLinkBox(target: target, link: link))
}

@_cdecl("ca_quartz_display_link_add_to_main_run_loop")
public func ca_quartz_display_link_add_to_main_run_loop(_ handle: UnsafeMutableRawPointer?) {
    guard #available(macOS 14.0, *), Thread.isMainThread, let box = caBorrowQuartzDisplayLinkBox(handle) else { return }
    box.link.add(to: .main, forMode: .default)
}

@_cdecl("ca_quartz_display_link_remove_from_main_run_loop")
public func ca_quartz_display_link_remove_from_main_run_loop(_ handle: UnsafeMutableRawPointer?) {
    guard #available(macOS 14.0, *), Thread.isMainThread, let box = caBorrowQuartzDisplayLinkBox(handle) else { return }
    box.link.remove(from: .main, forMode: .default)
}

@_cdecl("ca_quartz_display_link_invalidate")
public func ca_quartz_display_link_invalidate(_ handle: UnsafeMutableRawPointer?) {
    guard #available(macOS 14.0, *), let box = caBorrowQuartzDisplayLinkBox(handle) else { return }
    box.link.invalidate()
}

@_cdecl("ca_quartz_display_link_is_paused")
public func ca_quartz_display_link_is_paused(_ handle: UnsafeMutableRawPointer?) -> Bool {
    guard #available(macOS 14.0, *), let box = caBorrowQuartzDisplayLinkBox(handle) else { return false }
    return box.link.isPaused
}

@_cdecl("ca_quartz_display_link_set_paused")
public func ca_quartz_display_link_set_paused(_ handle: UnsafeMutableRawPointer?, _ paused: Bool) {
    guard #available(macOS 14.0, *), let box = caBorrowQuartzDisplayLinkBox(handle) else { return }
    box.link.isPaused = paused
}

@_cdecl("ca_quartz_display_link_get_timestamp")
public func ca_quartz_display_link_get_timestamp(_ handle: UnsafeMutableRawPointer?) -> Double {
    guard #available(macOS 14.0, *), let box = caBorrowQuartzDisplayLinkBox(handle) else { return 0 }
    return box.link.timestamp
}

@_cdecl("ca_quartz_display_link_get_duration")
public func ca_quartz_display_link_get_duration(_ handle: UnsafeMutableRawPointer?) -> Double {
    guard #available(macOS 14.0, *), let box = caBorrowQuartzDisplayLinkBox(handle) else { return 0 }
    return box.link.duration
}

@_cdecl("ca_quartz_display_link_get_target_timestamp")
public func ca_quartz_display_link_get_target_timestamp(_ handle: UnsafeMutableRawPointer?) -> Double {
    guard #available(macOS 14.0, *), let box = caBorrowQuartzDisplayLinkBox(handle) else { return 0 }
    return box.link.targetTimestamp
}

File: src/ca_display_link.rs
use core::ffi::c_void;

#[derive(Debug)]
/// Safe wrapper around `CADisplayLink`. See <https://developer.apple.com/documentation/quartzcore/cadisplaylink>.
pub struct QuartzDisplayLink {
    ptr: *mut c_void,
}

impl QuartzDisplayLink {
    #[must_use]
    /// Creates a `CADisplayLink` wrapper for the main screen.
    pub fn new_main_screen() -> Option<Self> {
        let ptr = unsafe { crate::ffi::ca_quartz_display_link_new_main_screen() };
        if ptr.is_null() {
            None
        } else {
            Some(Self { ptr })
        }
    }

    /// Adds the display link to the main run loop.
    pub fn add_to_main_run_loop(&self) {
        unsafe { crate::ffi::ca_quartz_display_link_add_to_main_run_loop(self.ptr) };
    }

    /// Removes the display link from the main run loop.
    pub fn remove_from_main_run_loop(&self) {
        unsafe { crate::ffi::ca_quartz_display_link_remove_from_main_run_loop(self.ptr) };
    }

    /// Invalidates the display link.
    pub fn invalidate(&self) {
        unsafe { crate::ffi::ca_quartz_display_link_invalidate(self.ptr) };
    }

    #[must_use]
    /// Returns whether the display link is paused.
    pub fn is_paused(&self) -> bool {
        unsafe { crate::ffi::ca_quartz_display_link_is_paused(self.ptr) }
    }

    /// Pauses or resumes the display link.
    pub fn set_paused(&self, paused: bool) {
        unsafe { crate::ffi::ca_quartz_display_link_set_paused(self.ptr, paused) };
    }

    #[must_use]
    /// Returns the display link timestamp.
    pub fn timestamp(&self) -> f64 {
        unsafe { crate::ffi::ca_quartz_display_link_get_timestamp(self.ptr) }
    }

    #[must_use]
    /// Returns the display link frame duration.
    pub fn duration(&self) -> f64 {
        unsafe { crate::ffi::ca_quartz_display_link_get_duration(self.ptr) }
    }

    #[must_use]
    /// Returns the display link target timestamp.
    pub fn target_timestamp(&self) -> f64 {
        unsafe { crate::ffi::ca_quartz_display_link_get_target_timestamp(self.ptr) }
    }
}

impl Drop for QuartzDisplayLink {
    fn drop(&mut self) {
        if !self.ptr.is_null() {
            unsafe {
                crate::ffi::ca_quartz_display_link_invalidate(self.ptr);
                crate::ffi::ca_release(self.ptr);
            }
            self.ptr = core::ptr::null_mut();
        }
    }
}
